drivers/gpio/gpio_mcp23008.c

```diff
@@ -0,0 +1,921 @@
+ *
+ * @param dev Device struct of the MCP23008
+ * @param mask Mask indicating which pins will be modified.
+ * @param value Value to set (0 or 1)
+ * @param toggle Mask indicating which pins will be toggled.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing the external GPIO chip over I2C.
+ * @retval -EWOULDBLOCK if operation would block.
+ */
+static int gpio_mcp23008_port_write(struct device *dev,
+				    gpio_port_pins_t mask,
+				    gpio_port_value_t value,
+				    gpio_port_value_t toggle)
+{
+	const struct gpio_mcp23008_config *const cfg = dev->config_info;
+	struct gpio_mcp23008_drv_data *const drv_data = dev->driver_data;
+	u8_t *pin_data = &drv_data->pin_state.data;
+	int rc = 0;
+	u8_t pin_data_old, reg_gpio;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		rc = -EWOULDBLOCK;
+		goto exitns;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/*
+	 * Only write changes to pin_state struct,
+	 * if they are successfully written to the device.
+	 */
+	pin_data_old = *pin_data;
+	reg_gpio = (u8_t) (((pin_data_old & ~mask) | (value & mask)) ^ toggle);
+
+	rc = (i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				 MCP23008_REG_GPIO, reg_gpio));
+	if (rc) {
+		LOG_ERR("Failed writing to i2c");
+		goto exit;
+	}
+
+	*pin_data = reg_gpio;
+
+exit:
+	k_sem_give(&drv_data->lock);
+exitns:
+	if (rc) {
+		LOG_ERR("%s: Could not write to port: %d",
+			dev->name, rc);
+	} else {
+		LOG_INF("%s: Wrote val 0x%02x msk 0x%02x: 0x%02x => 0x%02x",
+			dev->name, value, mask,
+			pin_data_old, *pin_data);
+	}
+	return rc;
+}
+
+/**
+ * @brief Non-ISR part for handling interrupt on IRQ line
+ * @param item Pointer to k_work structure
+ */
+static void gpio_mcp23008_isr_work_handler(struct k_work *item)
+{
+	struct gpio_mcp23008_drv_data *const drv_data =
+		CONTAINER_OF(item, struct gpio_mcp23008_drv_data, work);
+	struct device *const dev = drv_data->device_struct;
+	const struct gpio_mcp23008_config *cfg = dev->config_info;
+	u8_t intf, intcap;
+	int rc = 0;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		goto exitns;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	LOG_DBG("Read INTF");
+	rc = (i2c_reg_read_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				MCP23008_REG_INTF, &intf));
+	if (rc) {
+		LOG_ERR("Failed reading INTF");
+		goto exit;
+	}
+
+	/*
+	 * Reset INT line by reading out intcap register
+	 * (only edge triggered irqs)
+	 */
+	LOG_DBG("Read INTCAP");
+	rc = (i2c_reg_read_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				MCP23008_REG_INTCAP, &intcap));
+	if (rc) {
+		LOG_ERR("Failed reading INTCAP");
+		goto exit;
+	}
+
+	intf &= drv_data->pin_state.irq_enabled;
+
+	k_sem_give(&drv_data->lock);
+	gpio_fire_callbacks(&drv_data->cbs, dev, intf);
+
+	/*
+	 * Re-enable Interrupt since it was disabled in ISR.
+	 * We can assume that the interrupt was configured,
+	 * because else the isr could have never been called
```

Comment:
You can assume that, however you can't assume that a callback didn't disable the interrupt.  You need to check for that.

Related: I don't think the gpio_basic_api test would pass in this case, since it expects to be able to disable interrupts.  Please make sure both gpio driver API tests (1-pin and 2-pin) are passing, and note that in a comment on the PR.

Reply:
I'm not sure whether I understand.
We are talking about the interrupt of the mcp23008 int-line here right?
That one is statically enabled at the driver initialization and in theory continuously on.
The callbacks of this driver have no control over that interrupt, only callbacks on the driver of the gpio-controller attached to the int-line. 
As far as I know, we have no possibility to see whether a different driver/application disabled the interrupts on that gpio-controller while they were already disabled by us.

Even if it was possible, what would be the expected behavior of this driver then?
It would be configured to react on interrupts (no polling), but the int line is then disabled by another driver/app implicitly.

Maybe I'm just misunderstanding something.